tests/gpflow/experimental/check_shapes/test_error_contexts.py

```diff
@@ -0,0 +1,575 @@
+                    TestContext("C"),
+                    TestContext("D"),
+                ),
+            ),
+            """
+A
+  B
+    C
+      D
+""",
+        ),
+        (
+            StackContext(
+                StackContext(
+                    StackContext(
+                        TestContext("A"),
+                        TestContext("B"),
+                    ),
+                    TestContext("C"),
+                ),
+                TestContext("D"),
+            ),
+            """
+A
+  B
+    C
+      D
+""",
+        ),
+    ],
+)
+def test_stack_context(context: ErrorContext, expected: str) -> None:
+    assert expected == to_str(context)
+
+
+@pytest.mark.parametrize(
+    "context,expected",
+    [
+        (
+            ParallelContext([]),
+            """
+""",
+        ),
+        (
+            ParallelContext(
+                [
+                    TestContext("A"),
+                    TestContext("B"),
+                ]
+            ),
+            """
+A
+B
+""",
+        ),
+        (
+            ParallelContext(
+                [
+                    TestContext("A"),
+                    ParallelContext(
+                        [
+                            TestContext("B"),
+                            TestContext("C"),
+                        ]
+                    ),
+                    TestContext("D"),
+                    ParallelContext(
+                        [
+                            TestContext("E"),
+                            TestContext("F"),
+                        ]
+                    ),
+                    TestContext("G"),
+                ]
+            ),
+            """
+A
+B
+C
+D
+E
+F
+G
+""",
+        ),
+        (
+            ParallelContext(
+                [
+                    StackContext(
+                        TestContext("A"),
+                        StackContext(
+                            TestContext("B1"),
+                            TestContext("C1"),
+                        ),
+                    ),
+                    StackContext(
+                        StackContext(
+                            TestContext("A"),
+                            TestContext("B1"),
+                        ),
+                        TestContext("C2"),
+                    ),
+                    StackContext(
+                        TestContext("A"),
+                        StackContext(
+                            TestContext("B2"),
+                            TestContext("C1"),
+                        ),
+                    ),
+                ]
+            ),
+            """
+A
+  B1
+    C1
+    C2
+  B2
+    C1
+""",
+        ),
+    ],
+)
+def test_parallel_context(context: ErrorContext, expected: str) -> None:
+    assert expected == to_str(context)
+
+
+def test_function_call_context() -> None:
+    def f() -> str:
+        return to_str(FunctionCallContext(f))
+
+    assert (
+        f"""
+f called at: {__file__}:305
```

Comment:
(again you may be able to extract 305 using `getframeinfo(currentframe()).lineno`)

Reply:
> (again you may be able to extract 305 using `getframeinfo(currentframe()).lineno`)

Again, I don't really like the idea of `currentframe`, because that is (mostly) the same as what we're doing in the production code.